scheduler/preemption.go

```diff
@@ -0,0 +1,572 @@
+	return filteredBestAllocs
+
+}
+
+// PreemptForNetwork tries to find allocations to preempt to meet network resources.
+// This is called once per task when assigning a network to the task. While finding allocations
+// to preempt, this only considers allocations that share the same network device
+func (p *Preemptor) PreemptForNetwork(networkResourceAsk *structs.NetworkResource, netIdx *structs.NetworkIndex) []*structs.Allocation {
+
+	// Early return if there are no current allocs
+	if len(p.currentAllocs) == 0 {
+		return nil
+	}
+
+	deviceToAllocs := make(map[string][]*structs.Allocation)
+	MbitsNeeded := networkResourceAsk.MBits
+	reservedPortsNeeded := networkResourceAsk.ReservedPorts
+
+	// Create a map from each device to allocs
+	// We do this because to place a task we have to be able to
+	// preempt allocations that are using the same device.
+
+	// This step also filters out high priority allocations and allocations
+	// that are not using any network resources
+	for _, alloc := range p.currentAllocs {
+		if alloc.Job == nil {
+			continue
+		}
+
+		if p.jobPriority-alloc.Job.Priority < 10 {
+			continue
+		}
+		networks := alloc.ComparableResources().Flattened.Networks
+
+		// Only include if the alloc has a network device
+		if len(networks) > 0 {
+			device := networks[0].Device
+			allocsForDevice := deviceToAllocs[device]
+			allocsForDevice = append(allocsForDevice, alloc)
+			deviceToAllocs[device] = allocsForDevice
+		}
+	}
+
+	// If no existing allocations use network resources, return early
+	if len(deviceToAllocs) == 0 {
+		return nil
+	}
+
+	var allocsToPreempt []*structs.Allocation
+
+	met := false
+	freeBandwidth := 0
+
+	for device, currentAllocs := range deviceToAllocs {
+		totalBandwidth := netIdx.AvailBandwidth[device]
+
+		// If the device doesn't have enough total available bandwidth, skip
+		if totalBandwidth < MbitsNeeded {
+			continue
+		}
+
+		// Track how much existing free bandwidth we have before preemption
+		freeBandwidth = totalBandwidth - netIdx.UsedBandwidth[device]
+
+		preemptedBandwidth := 0
+
+		// Reset allocsToPreempt since we don't want to preempt across devices for the same task
+		allocsToPreempt = nil
+
+		// Build map from used reserved ports to allocation
+		usedPortToAlloc := make(map[int]*structs.Allocation)
+
+		// First try to satisfy needed reserved ports
+		if len(reservedPortsNeeded) > 0 {
+			for _, alloc := range currentAllocs {
+				for _, tr := range alloc.TaskResources {
+					reservedPorts := tr.Networks[0].ReservedPorts
+					for _, p := range reservedPorts {
+						usedPortToAlloc[p.Value] = alloc
+					}
+				}
+			}
+
+			// Look for allocs that are using reserved ports needed
+			for _, port := range reservedPortsNeeded {
+				alloc, ok := usedPortToAlloc[port.Value]
+				if ok {
+					preemptedBandwidth += alloc.Resources.Networks[0].MBits
+					allocsToPreempt = append(allocsToPreempt, alloc)
+				}
```

Comment:
If !ok can't you continue on the outer for loop since there is a port that can't be acquired (alloc is too high of priority)

Reply:
I'll add a test case for this, good catch